### PR TITLE
Refactor matrix construction and evaluation

### PR DESCRIPTION
**Pull request by huisaddison**
This changes the C++ routines to return `Eigen::SparseMatrix<double>` rather than List of Vectors.  There is no change to the R interface.  The conversion of `Eigen::SparseMatrix<double>` to `dgCMatrix` is performed transparently by `RcppEigen`.

Advantages:

* Other R packages which link to dspline C++ functions can now receive sparse matrices directly from function calls, rather than having to receive a List and manually construct the sparse matrix.  This is the biggest advantage.
* The C++ interface and the division of labor in code itself is simpler.  Previously, the code would construct `i_vec, j_vec, x_vec` in C++ and have R turn these vectors into a sparse matrix (via a separate call to C++ code in the Matrix package) after doing some input verification.  Now the R code just does input verification, and the C++ code does everything else.

Notes:

* Just a matter of "minimalism" in the code: the `matrix_[construction, evaluation].cpp` additionally "depends" on Eigen whereas previously it only depended on Rcpp.  But I think this is a worthwhile tradeoff to allow other packages to receive sparse matrices directly in C++.
* ~~The only part of the refactor I'm not fully satisfied with is in `rcpp_n_mat`.  The use of `i_vec, j_vec, x_vec` cannot be eliminated easily since the construction uses the function `nj` that operates in-place on these vectors.  So the code isn't as elegant but I still don't think it's any more efficient than before (passing vectors to R to be converted to sparse matrices).~~ This is refactored in later commits.

### CODE REVIEW

**Comment by huisaddison** [comment]
All tests pass.  Documentation unchanged since the R interface has not changed.

```
$ ./make -t
*** Compiling ... ***
*** Loading ... ***
ℹ Updating dspline documentation
Setting `RoxygenNote` to "7.2.3"
ℹ Loading dspline
*** Testing ... ***
ℹ Testing dspline
✔ | F W S  OK | Context
✔ |        51 | dspline [0.5s]

══ Results ══════════════════════════════════
Duration: 0.5 s

[ FAIL 0 | WARN 0 | SKIP 0 | PASS 51 ]
```

**Comment by huisaddison** [comment]
I'm converting this to a draft because we can improve the refactor of `rcpp_n_mat` by also refactoring `nj`.  In short, rather than declaring and updating `i_vec, j_vec, x_vec`, just declare and update `n_list`.  This also eliminates the need to keep a global `l` index.

Note to self: I don't understand how, as currently written, the code is passing `i_vec`, etc., by reference.  I would have thought that the function declaration requires the types to be annotated with an ampersand to specify that those arguments are passed by reference, not value.

**Comment by ryantibs** [comment]
Let's go through this together next Thursday.
Thanks and great work thus far 👏 !!

**Comment by huisaddison** [comment]
Sounds good!  I did have a chance to read about what Rcpp does under the hood (the proxy model), and I comprehend the reference-like behavior of the Rcpp Vector classes now.

We should talk about it more on Thursday, but my basic thoughts on best practices:
* Rcpp classes should be used only if an object with interact with R (input, output, or in-place modification of an existing R object)
* Eigen and std classes should be preferred otherwise: the code is more portable, we have better control of pass by value vs reference, and likely we would need to convert into one of those classes anyways (e.g., for linear algebra).

I also wanted to note the following behavior so we make sure to discuss it.  The matrix construction and evaluation functions return `dgCMatrix` objects, even though the `Matrix` package hasn't been explicitly attached.  (Recall that this is possible because the `dgCMatrix` class is available, by way of importing the class into the dspline namespace via `DESCRIPTION`.  And previously, this was possible because the call to `Matrix::SparseMatrix` implicitly loads the `Matrix` namespace; see the documentation for `::`.)  But, if `Matrix` package has been explicitly attached, the result of `solve(SparseMatrix, Vector)` is a `dgeMatrix` single-column dense matrix; else it is (I assume coerced into) a numeric vector.  (Note that this behavior predates this refactor since `Matrix` was not explicitly attached with `dspline` either.)

My hope is that our code is "correct" as written, and that this the "expected behavior" of `solve`, but I would like to see documentation that verifies this.  We should also discuss whether we want to have dspline `Depends: Matrix`, which forces `Matrix` to be attached along with `dspline`.  If so, a user would always have the former behavior, and there would be no confusion.

```
> library(dspline)
> set.seed(1)
> n = 10
> k = 2
> xd = sort(runif(n))
> v = rnorm(n)
> solve(b_mat(k, xd), v)
 [1] -0.8204684 -0.7522792 -0.7163680 -0.6511510 -0.5377841 -0.4950805
 [7] -0.4704884 -0.3059487 -0.3018392 -0.2856259
> head(methods(solve))
[1] "solve,ANY,ANY-method"
[2] "solve,ANY,Matrix-method"
[3] "solve,CHMfactor,ANY-method"
[4] "solve,CHMfactor,ddenseMatrix-method"
[5] "solve,CHMfactor,diagonalMatrix-method"
[6] "solve,CHMfactor,dsparseMatrix-method"
> library(Matrix)
> solve(b_mat(k, xd), v)
10 x 1 Matrix of class "dgeMatrix"
            [,1]
 [1,] -0.8204684
 [2,] -0.7522792
 [3,] -0.7163680
 [4,] -0.6511510
 [5,] -0.5377841
 [6,] -0.4950805
 [7,] -0.4704884
 [8,] -0.3059487
 [9,] -0.3018392
[10,] -0.2856259
```

File: src/matrix_construction.cpp [hunk 1]
```diff
@@ -130,18 +144,65 @@ List rcpp_h_eval(int k, NumericVector xd, NumericVector x, IntegerVector col_idx
...
+// Evaluation of discrete spline basis at arbitrary query points
+
+// [[Rcpp::export]]
+Eigen::SparseMatrix<double> rcpp_n_eval(int k, NumericVector xd, NumericVector x, bool normalized, IntegerVector knot_idx) {
+  int n_xd = xd.size();
+  int n_knot_idx = knot_idx.size();
+  double max_diff = Rcpp::max(xd[Rcpp::seq(1, n_xd-1)]-xd[Rcpp::seq(0, n_xd-2)]);
+  IntegerVector ext_knot_idx(n_knot_idx+k+1);
+  NumericVector ext_xd(n_xd+k+1);
+  ext_knot_idx[Rcpp::seq(0, n_knot_idx-1)] = knot_idx;
```
Comment (huisaddison) [inline src/matrix_construction.cpp:172]: Should we move this to always occur inside `rcpp_n_mat` and remove the corresponding code from the R `n_mat` function?  Would anybody ever want to call `rcpp_n_mat` _without_ the extra knots appended?

File: src/matrix_construction.cpp [hunk 2]
```diff
@@ -130,18 +156,62 @@ List rcpp_h_eval(int k, NumericVector xd, NumericVector x, IntegerVector col_idx
...
+// Evaluation of discrete spline basis at arbitrary query points
+
+// [[Rcpp::export]]
+Eigen::SparseMatrix<double> rcpp_n_eval(int k, NumericVector xd, NumericVector x, bool normalized, IntegerVector knot_idx) {
+  int n = xd.size();
+  int n_knot_idx = knot_idx.size();
+  double max_diff = Rcpp::max(xd[Rcpp::seq(1, n-1)]-xd[Rcpp::seq(0, n-2)]);
+  IntegerVector ext_knot_idx = concat<IntegerVector>(knot_idx,
+      Rcpp::seq(n-1, n+k-1));
+  NumericVector extra_knots = Rcpp::cumsum(Rcpp::rep(max_diff, k+1));
```
Comment (huisaddison) [inline src/matrix_construction.cpp:184]: Should we move this to always occur inside `rcpp_n_mat` and remove the corresponding code from the R `n_mat` function? Would anybody ever want to call `rcpp_n_mat` without the extra knots appended?

**Comment by huisaddison** [comment]
@ryantibs I added the changes from [this commit](https://github.com/ryantibs/dspline/commit/041546dfb2550a450be5a698b772dfb36423df5b).  

The instructions given by Github essentially created a branch on your repository with the name `huisaddison-refactor/matrix-construction` and played my commits onto them (because you don't have write access to my fork).  Then you made the above commit on top of the branch in _your_ repository.  (For some reason, it combined the last commit [ba9378e](https://github.com/ryantibs/dspline/pull/9/commits/ba9378e0f663299b33286822a2d874c776143a26) into your "Tiny changes" commit).

Anyways, I cannot write to that branch, so I figured it'd be faster for me to make the change myself rather than to ask you to re-make the change on this branch.  In principle, though, you should have been able to write to my branch since I issued this PR with ["Allow edits by maintainer" turned on](https://github.blog/2016-09-07-improving-collaboration-with-forks/).  Maybe there is a special way to set upstream in order for you to write to my branch, but you instead wrote to a "regular" branch?

**Comment by ryantibs** [comment]
Thanks! Weird about the commits thing.

Question: should I merge this (after running tests), and we make a new issue for use of `Rcpp::seq()` versus `Range()`, use of `Rcpp::max()` instead of `std::something(), and any other related things like that? 

Or instead, should we address that in this PR?

**Comment by huisaddison** [comment]
All tests pass:

```
*** Compiling ... ***
*** Loading ... ***
ℹ Updating dspline documentation
ℹ Loading dspline
*** Testing ... ***
ℹ Testing dspline
✔ | F W S  OK | Context
✔ |        51 | dspline [0.5s]

══ Results ═══════════════
Duration: 0.5 s

[ FAIL 0 | WARN 0 | SKIP 0 | PASS 51 ]
```

**Comment by ryantibs** [comment]
Way to---great changes!!

**Pull request merged**